### ISSUE TITLE
How to operate on a layer within a layered spec?

### ISSUE THREAD

**Issue by ijlyttle** [vegawidget/vlbuildr#13]
Hi @AliciaSchep,

As you may know, **ggvega**, is starting to be functional; we can make some scatterplots.

I've been putting together the UseR! presentation (I need to send you the slides to have a look at) - I show how you can build an initial spec using **ggplot2**, then **ggvega**, *then* modify it using **vlbuildr**, and it works. 

I ran into a problem, though, when trying to add a selection to a layered spec (by design, ggvega returns a layered spec). 

Forgive me if there is a workaround already and I did not see it, but I found a workaround that you might find interesting. I put it in a notebook here https://rstudio.cloud/project/398318 (it's in a file called `ggvega-vlbuildr.Rmd`).

Tagging in @haleyjeppson, in case she might be interested.

Thanks!

**Comment by AliciaSchep** [comment]
Yes, at the moment adding a selection can't be done on a layer spec.  (As you know) a selection isn't a valid specification at the top-level of a layer spec. If the current functions to add selections were to operate on a layer spec, the question would be which of the inner layer to apply the selection to.  One reasonable answer to that question might be (for both selection and any other similar functions that operate at the same level) to allow such functions when the layer only has 1 element and there is no choice to make. Another option could be to add to the last  in the list of layers.

In terms of how `vlbuildr` is now,, when making composite plots you would generally need to build up each sub-component separately and then assemble, without much option to edit the sub-components when assembled.  I think your idea of a `with_layer` is really interesting.  I have been thinking of whether the api should be more flexible in terms of being able to build on layered or composite plots.  Another option would be to for most of the functions to add S3 methods for the different types of specs; if it is a layer spec it has an extra argument (`.layer`) which might default to the last layer but can be set by the user. That might enable building up a layer plot like so:

```r
vl_chart() %>%
  add_some_stuff for first layer...
  vl_add_layer() %>% # a function to initialize an additional layer
  add_some_stuff to second layer...
  add something to the first layer using .layer argument...
```

Where this would get complicated is that a `layer` plot can have a `LayerSpec` as one of the internal specs... so after you have chosen the right 'layer' at the first level, how do you choose the layer within the second level?

**Comment by ijlyttle** [comment]
I think we are thinking along the same lines here; that *building* a spec and *performing surgery* on a spec are two different things. 

I know I am bring lazy by not looking for it now (so maybe I'll edit this later), does **ggplot2** offer any tools for "plot-surgery", and if so, what is its API?

I have modified the notebook so that `with_layer()` defaults to the last layer; it's new signature is:

```r
with_layer <- function(.f, .layer = NULL)
``` 

On your question about nested layers, could the `.layer` argument permit a non-scalar, e.g. `.layer = c(2, 3)`, and its default could be the last of the last of the last...? 

Perhaps the `with_layer()` approach is more verbose for users (and removes auto-complete), but keeps your functions "cleaner". Of course, by definition, you have the best view of this.

**Comment by ijlyttle** [comment]
Even crazier idea here - I have no idea if it can be made to work, or even if it is helpful. From the user's perspective, how about something like:

```r
vl_petal <-
  vl_petal %>%
  on_layer({
    vl_add_interval_selection("brush") %>%
    vl_encode_opacity(value = 0.3) %>%
    vl_condition_opacity(selection = "brush", value = 1)
  })
```

with the signature:

```r 
on_layer <- function(expr, .layer = NULL)
```

**Comment by AliciaSchep** [comment]
I like that idea! I wonder if it should also apply to `concat` specs as well?  E.g maybe something more generic like `on_level` or `at_level` that can apply to `concat` or `layer` specs?

**Comment by ijlyttle** [comment]
For the purpose of showing something at UseR!, I have hacked together something at the [RStudio Cloud project](https://rstudio.cloud/project/398318). In summary, it would suggest a syntax like this:

```r
vl_at_last_layer <- function(spec, code)
```

```r
vl_petal <-
  vl_petal %>%
  vl_at_last_layer({
    . %>%
    vl_add_interval_selection("brush") %>%
    vl_encode_opacity(value = 0.3) %>%
    vl_condition_opacity(selection = "brush", value = 1)
  })
```

A couple of notes on this idea:

- I note that this is **experimental** syntax
- adding the `vl_` prefix to match the other functions
- the only way I have been able to get the evaluation to work is to start the code with `. %>%` (thanks to @jimhester for pointing this out!)
- the function `vl_at_last_layer()`  (maybe `vl_at_last()` ?) could be a wrapper to a more general function:

```r
vl_at_level <- function(spec, level, code) 
```

I guess my question (esp for @AliciaSchep) is to ask if including this experimental syntax for `vl_at_last_layer()` would be OK to show in a demo. 

Of course, there is nothing set in stone about this demo, including the function names - and it's not until Thursday. For the demo, I could call even call the function something like `this_function_will_likely_change()`.



Thanks!

**Comment by AliciaSchep** [comment]
@ijlyttle feel free to show in demo as long as noted as experimental & subject to change!  Perhaps a different color font or something like that could be used to demarcate experimental part (although the whole package could likely still be considered experimental though!)

Some further thoughts... I am trying to think through pros/cons of various approaches for the core issue of being able to modify an inner spec, including:

(1) the approach you propose or something similar (where you put expressions in a function that says what layer to operate).  One nice features of this approach is that it is very readable to see the group of functions being applied at a given level.  Having to start off the expression with `. %>%` is a bit clunky though...

(2) making most `vl_* `  functions accept a 'level' argument to indicate what level to apply the function.  A disadvantage here is that if you were to do your example above you would need to specify the level multiple times (unless using default level). An advantage is that if using the default level (probably last) then the use is pretty effortless (although that may be a disadvantage too as it is no longer explicit).

(3) Something else entirely.  One idea is to give a spec optionally an attribute that holds 'current level'.  Then any function applied will use that level. Something like this: 

```r
vl_petal <-
  vl_petal %>%
  vl_at_last_layer() %>% # Set level attribute so subsquent function operate on inner spec
  vl_add_interval_selection("brush") %>%
  vl_encode_opacity(value = 0.3) %>%
  vl_condition_opacity(selection = "brush", value = 1) %>%
  vl_at_top_level() # return to top level; not needed here but could be if wanting to add things to top level
```

**Comment by ijlyttle** [comment]
Hi @AliciaSchep - just a quick few (late) words to say I agree that the `. %>%` does not "roll off the tongue", so looking forward to this discussion. I think I need to use it for the demo, but there will be no mistaking this for anything official :)

At this point, I find the "`# Set level attribute ...`" idea very interesting!

**Comment by ijlyttle** [comment]
Acting as @AliciaSchep: Look at how 'here' handles relative paths (maybe can use the same to get at relative places in the spec)

**Comment by ijlyttle** [comment]
Acting for me - think about "ggvega" examples where we would want to add selections, conditional encoding, etc.
